module-apps/apps-common/ApplicationCommon.cpp

```diff
@@ -531,63 +543,81 @@ namespace app
+        if (switchData && switchData->ignoreCurrentWindowOnStack) {
+            windowsStack().pop();
+        }
+        auto anotherWindowOnTop = (not isCurrentWindow(msg->getWindowName())) and (not windowsStack().isEmpty());
+        if (anotherWindowOnTop) {
+             auto closeReason = gui::Window::CloseReason::WindowSwitch;
+             switch (msg->getReason()) {
+             case SwitchReason::PhoneLock:
+                 closeReason = gui::Window::CloseReason::PhoneLock;
+                 break;
+             case SwitchReason::Popup:
+                 closeReason = gui::Window::CloseReason::Popup;
+                 break;
+             default:
+                 break;
+             }
+            getCurrentWindow()->onClose(closeReason);
+        }
+
+        LOG_DEBUG("Current window: %s vs %s", getCurrentWindow()->getName().c_str(), msg->getWindowName().c_str());
+        // THIS IS IN TEST!
+        const auto &[name, data] = msg->getSwitchData();
+        pushWindow(name, data);
+        getCurrentWindow()->handleSwitchData(switchData.get());
+
+        if (handleUpdateTextRefresh(switchData.get())) {
+            return sys::msgHandled();
+        }
+        getCurrentWindow()->onBeforeShow(msg->getCommand(), switchData.get());
 
-            auto ret = dynamic_cast<gui::SwitchSpecialChar *>(switchData.get());
-            if (ret != nullptr && ret->type == gui::SwitchSpecialChar::Type::Response) {
-                auto text = dynamic_cast<gui::Text *>(getCurrentWindow()->getFocusItem());
-                if (text != nullptr) {
-                    text->addText(ret->getDescription());
-                    refreshWindow(gui::RefreshModes::GUI_REFRESH_FAST);
-                    return sys::msgHandled();
-                }
-            }
-            getCurrentWindow()->onBeforeShow(msg->getCommand(), switchData.get());
+        /// This is kind of gentle hack on our stateless Application Window management switching
+        /// if we are not requesting for popup and we cant handle next popup - just refresh the window
+        /// else -> popup will refresh it for us
+        if (not tryShowPopup()) {
             refreshWindow(gui::RefreshModes::GUI_REFRESH_DEEP);
         }
-        else {
-            LOG_ERROR("No such window: %s", msg->getWindowName().c_str());
-        }
         return sys::msgHandled();
     }
 
+    bool ApplicationCommon::handleUpdateTextRefresh(gui::SwitchData *data)
+    {
+        auto ret = dynamic_cast<gui::SwitchSpecialChar *>(data);
+        if (ret != nullptr && ret->type == gui::SwitchSpecialChar::Type::Response) {
+            auto text = dynamic_cast<gui::Text *>(getCurrentWindow()->getFocusItem());
+            if (text != nullptr) {
+                text->addText(ret->getDescription());
+                refreshWindow(gui::RefreshModes::GUI_REFRESH_FAST);
+                return true;
+            }
+        }
+        return false;
+    }
```

Comment:
What is the purpose of this?

Reply:
this was moved out from generic switchWindow handling and is a tech debt :) though its name should be self-explanatory?